cranelift-codegen/src/isa/x86/enc_tables.rs

```diff
@@ -893,3 +894,63 @@ fn expand_fcvt_to_uint_sat(
     cfg.recompute_ebb(pos.func, uint_large_ebb);
     cfg.recompute_ebb(pos.func, done);
 }
+
+/// Because floats already exist in XMM registers, we can keep them there when executing a CLIF
+/// extractlane instruction
+fn convert_extractlane(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::ExtractLane {
+        opcode: ir::Opcode::Extractlane,
+        arg,
+        lane,
+    } = pos.func.dfg[inst]
+    {
+        // remove the original extractlane instruction
+        pos.remove_inst();
+
+        // extract the original result value and use it (e.g. with_result) as the final result value of our replaced sequences
+        let original_results = pos.func.dfg.detach_results(inst);
+        let original_result = original_results.get(0, &pos.func.dfg.value_lists).unwrap();
+
+        // NOTE: the following legalization assumes that the upper bits of the XMM register do not need to be zeroed during extractlane
+        let value_type = pos.func.dfg.value_type(arg);
+        if value_type.lane_type().is_float() {
+            // floats are already in XMM registers and can stay there
+            if lane != 0 {
+                // we replace the extractlane with a PSHUFD to get the float in the right place
+                let shuffled = match value_type {
+                    F32X4 => {
+                        let shuffle_mask: u8 = 0b00_00_00_00 | lane; // we move the selected lane to the 0 lane
+                        pos.ins().x86_pshufd(arg, shuffle_mask)
+                    }
+                    F64X2 => {
+                        assert_eq!(lane, 1);
+                        let shuffle_mask = 0b11_10_11_10; // because we know the lane == 1, we move the upper 64 bits to the lower 64 bits, leaving the top 64 bits as-is
+                        let bitcast = pos.ins().raw_bitcast(F32X4, arg);
+                        pos.ins().x86_pshufd(bitcast, shuffle_mask)
+                    }
+                    _ => unreachable!(),
+                };
+                // then we must bitcast to the right type
+                pos.ins()
+                    .with_result(original_result)
+                    .raw_bitcast(value_type.lane_type(), shuffled);
+            } else {
+                // remove the extractlane instruction, leaving the float where it is
+                pos.ins()
+                    .with_result(original_result)
+                    .raw_bitcast(value_type.lane_type(), arg);
```

Comment:
Maybe control flow could be slightly simplified here: both branches end up with a raw_bitcast, so the `if` could yield the raw_bitcast argument instead?